Docker image scans GH action

Nava-JoshLong opened the pull request:
## Ticket

Resolves #207 

## Changes
- Added GitHub action for scanning a Docker image for vulnerabilities to ensure compliance
- Added README to the action's directory so that it can be referenced when needed
- Added template file for skipping specified CVE

## Context for reviewers
This action allows for scanning of an image before it is pushed to ECR and deployed to ECS clusters, so if there are any vulnerabilities they are found and fails the CD process. ECR can be set to scan on push, but by the time the scan is finished, the image might already be deployed. 

## Testing
To test, add this step to a workflow, where the stage the creates the Docker image has the id `build-image` and appends the image to `$GITHUB_OUTPUT` (ie `echo "image=$ECR_REGISTRY/$ECR_REPOSITORY:$IMAGE_TAG" >> $GITHUB_OUTPUT`), or change the `image: ${{ steps.build-image.outputs.image }}` line to the image that is built 

```
      - name: Scan docker image
        uses: ./.github/actions/image-scan
        with:
          image: ${{ steps.build-image.outputs.image }}
```

lorenyu commented on .grype.yml:
```diff
@@ -0,0 +1,23 @@
+ignore:
```
Could we add a comment block at the top of the file explaining what this file is, what it is used for, and link to other relevant docs that say more (e.g. vulnerability-management.md)

Nava-JoshLong replied:
Yeap, I'm going to change how this is set in the next commit. I cleaned it up in our repo and will pull the changes over

lorenyu commented on .github/actions/image-scan/action.yml:
```diff
@@ -0,0 +1,26 @@
+    description: The docker image to scan
+    required: true
+runs:
+  using: "composite"
```
What's the value of making this a composite action instead of having the projects call trivy and anchore scans directly? As a point of comparison, [we call tfsec and checkov directly in ci-infra.yml](https://github.com/navapbc/template-infra/blob/main/.github/workflows/ci-infra.yml#L41-L67). One benefit I could see is that it hides away all the parameters. But the disadvantage is it's an extra layer of abstraction that I'm not 100% sure is needed.

Nava-JoshLong replied:
I think there can be an argument for both cases: 
 - By having it as a composite, you can call that from any pipeline, and expect the output to be the same, and don't need to update in many places when there are any changes you'd like to do
 - By using it directly, you don't have to bother with using composite actions, which can sometimes cause fits because of weird behavior

I can make a workflow that has it directly used instead of calling the action, and we can let the users decide which is best for their setup

lorenyu replied:
Can you say more about "By having it as a composite, you can call that from any pipeline, and expect the output to be the same, and don't need to update in many places when there are any changes you'd like to do"? What are some example use cases for scanning the image outside of CI?

Nava-JoshLong replied:
Meaning if we had the jobs in multiple places instead of a composite, any changes would have to be done to all of them to have the same output. Admittedly, that's not that hard to update it. 

As for outside of CI, in our pipelines right now, we scan in the CI and CD, but I think moving it to just CI is the better move

lorenyu commented on .github/actions/image-scan/action.yml:
```diff
@@ -0,0 +1,26 @@
+    required: true
+runs:
+  using: "composite"
+  steps:
```
If I understand correctly, these steps are currently set up to run synchronously, but they don't depend on each other so theoretically they could run in parallel as separate jobs. Is that correct?

lorenyu replied:
I think your example is describing what I'm saying – anchore-scan and trivy-scan need to be separate jobs in order to run in parallel. Putting them as separate steps in a single job or composite action will run them in serial. I don't know of a way to make the various steps in a `steps` section of a github action run in parallel.

lorenyu commented on docs/infra/vulnerability-management.md:
```diff
@@ -0,0 +1,31 @@
+# Vulnerability Management for Docker Images
+This repository contains a GitHub action that allows you to scan Docker images for vulnerabilities. The action, named `image-scan` is located in the directory `.github/actions/image-scan`. The goal in scanning the image before pushing it to the repository is so that you can catch any vulnerabilities before deploying the image, ECR scanning takes time and the image can still be used even with vuln ...
```
I'm wondering, when should these scans be run? I'm guessing they should be run on PRs. Should they also run on merges to main, or is that redundant?

Nava-JoshLong replied:
I'm wondering the same thing right now, I think PRs for sure - dependabot is trying to merge a PR that has a vulnerability for us right now. But, I think it should on a merge to main **IF** the last scan of the image is like X days before the merge to main. That way, in case there was something that was found between the last push to the PR and the merge, it should get caught

lorenyu replied:
That makes sense. In that case I think `on: pull_request` and `on: push: [main]` is probably good. No harm in running on every merge

lorenyu reviewed (commented):
This is awesome, great to see the PR and now that I'm seeing it concretely I have a much better understanding of it now.

The main thought I have is that instead of only including an action and therefore requiring project teams to put this into their workflows, it would be nice to go and add a template workflow that does the image scanning. Maybe something like `.github/workflows/ci-vulnerability-scans.yml` or something.

Thoughts?

The other thought is that I'm not 100% convinced we need the composite action. We could potentially just call the trivy and anchor scans directly in the ci-vulnerability-scans.yml workflow. The added advantage is that the jobs can be run in parallel.

lorenyu commented on .grype.yml:
```diff
@@ -0,0 +1,23 @@
+# https://github.com/anchore/grype#specifying-matches-to-ignore
+# More info can be found in the docs/infra/vulnerability-management.md file
+
+# Please add whitelists in the following format to make it easier when checking
```
Nit: Nava tends to use the word safelist instead of whitelist
```suggestion
# Please add safelists in the following format to make it easier when checking
```

Nava-JoshLong replied:
Gotcha, I'll have that in the next commit. I am going to make an example workflow that runs the scans in parallel

Nava-JoshLong commented:
Sorry for the long time between pushes, had PTO, jury duty and other things pop up in the time between. This commit has new scans added and some fixes on the documentation that you called out. Looks like I'll need to do some more tweaking to make it happy, which hopefully won't take as long. One thing I still was not able to figure out is how to share a Docker image between jobs so that we can run scans and make tests in parallel

Nava-JoshLong commented:
@lorenyu For the failing job, not too sure why the workflow isn't outputting nice, but locally I was getting the same errors. I wanted to show you it finding issues with the template Dockerfile. Here's the issues it's showing locally
```
[API Linting & Testing CI/image-scan]   ⚙  ::set-output:: results=app/Dockerfile:13 DL3007 warning: Using latest is prone to errors if the image will ever update. Pin the version explicitly to a release tag
app/Dockerfile:15 DL3020 error: Use COPY instead of ADD for files and folders
app/Dockerfile:16 DL3020 error: Use COPY instead of ADD for files and folders
app/Dockerfile:17 DL3020 error: Use COPY instead of ADD for files and folders
app/Dockerfile:20 DL3025 warning: Use arguments JSON notation for CMD and ENTRYPOINT arguments
```

We can cast some of these as lower levels, but I can work on resolving them as I have time if you want

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,81 @@
+        with:
+          image: ${{ steps.build-image.outputs.image }}
+
+  linting-and-testing:
+    runs-on: ubuntu-latest
+    needs: check-for-changes
+    if: ${{ needs.check-for-changes.outputs.needs-testing == 'true' }}
+    defaults:
+      run:
+        working-directory: ./app
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Initialize the docker containers
+        run: make init
+
+      # Change this step to your testing make commands, you can run a list here
+      - name: Run Tests
+        run: make testing
+
+  lint-test:
+    # Orchestration job that runs the other jobs in parallel
+    runs-on: ubuntu-latest
+    needs:
+      - check-for-changes
+      - linting-and-testing
+      - image-scan
+    if: ${{ needs.check-for-changes.outputs.needs-testing == 'true' }}
+    steps:
+      - run: echo "Finished running tests successfully!"
```
Let's remove application testing and linting. Those should be implemented in individual application repos in `ci-app.yml` (see https://github.com/navapbc/template-application-flask/blob/main/.github/workflows/ci-app.yml and https://github.com/navapbc/template-application-nextjs/blob/main/.github/workflows/ci-app.yml) and is testing the application code rather than the infrastructure

Nava-JoshLong replied:
Should I remove it here and make a template that has it to show the process?

lorenyu replied:
What did you have in mind? template-application-flask and template-application-nextjs both have it already, so anyone using the template application would already have an example application linting and testing. Were you thinking of a third template or something else?

Nava-JoshLong replied:
For a future PR, make a third one with them all in it. I was thinking if we are scanning an image that the linting and testing jobs might be running against, we could share the image so we're building it once. Obviously can't do that for this PR since we haven't been able to get it working. 

Also, to maybe make things somewhat cleaner by needing a single job compared to multiple jobs for merging. We have our repo rules to not allow merges from PRs until the orchestration job is successful, so having a single job call all the other jobs would make it somewhat easier than needing to have multiple files that have multiple blocking jobs required for merges

lorenyu replied:
Oh, in terms of example repos that integrate template-infra and template-application and have all of the CI jobs running, we have [platform-test](https://github.com/navapbc/platform-test) and [platform-test-nextjs](https://github.com/navapbc/platform-test-nextjs). I haven't yet created platform-test-flask but it's in the backlog.

lorenyu replied:
> Also, to maybe make things somewhat cleaner by needing a single job compared to multiple jobs for merging. We have our repo rules to not allow merges from PRs until the orchestration job is successful, so having a single job call all the other jobs would make it somewhat easier than needing to have multiple files that have multiple blocking jobs required for merges

I don't think putting all the checks in one job is cleaner. I think it's cleaner to have the checks separate, so that you have a sense of what is failing just by looking at which check failed, rather than having to click into the check.

Nava-JoshLong replied:
For you first comment, yes something like that. For the second one, I was meaning something along the lines of this

- Branch protection rule requires job named `image-scans` to be successful
- That job runs all other jobs in parallel (as in this action in the PR)
- Any fails will fail the overall job, where you can click into that (see below in the checks)
- Any new jobs that we might want to add in the future would just be added to the orchestration job to run

I'm saying this because in NJ, our infra is in a separate repo, so any changes like required checks need a different PR to do it

lorenyu replied:
Maybe I'm misunderstanding, but how is changing required checks related to infrastructure? Isn't that just Github setting that you'd change through the UI (see screenshot below)? Or are y'all managing github repo settings via code somehow?

<img width="784" alt="image" src="https://user-images.githubusercontent.com/447859/214186214-552b4470-6f27-4bec-a486-0c24e8ad3d63.png">

Nava-JoshLong replied:
Yeah, we managed the settings with Terraform. There's a couple things that were changed manually since I deployed the settings, and I need to update our statefile with that in case we ever need to do a redeploy or copy the settings to another repo for another one of our teams on NJ

lorenyu commented on app/Makefile:
```diff
@@ -7,3 +7,11 @@ release-build:
 		--platform=linux/amd64 \
 		$(OPTS) \
 		.
+
+init:
+	@echo "Example command for ci-vulnerability-scans.yml pipeline to build Docker image"
+	@echo "If you are using this file, update this command to build your app for testing"
+
+testing:
+	@echo "Example command for ci-vulnerability-scans.yml pipeline to run tests"
+	@echo "This is where any testing for your CI would go"
```
This is also covered by the application (see https://github.com/navapbc/template-infra/blob/main/template-only-docs/application-requirements.md).

and as an example see https://github.com/navapbc/template-application-nextjs/blob/main/app/Makefile and https://github.com/navapbc/template-application-flask/blob/main/app/Makefile which both define `make release-build` in `app/Makefile`, so we don't need it here.

lorenyu reviewed (commented):
Thanks for following up! No worries about the delay between changes.

The current PR still looks like it uses a composite action but we discussed in [this thread](https://github.com/navapbc/template-infra/pull/203/files#r1047497700) to simplify, get rid of the action, and just run the scanning jobs in parallel.

Nava-JoshLong commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,81 @@
+# This file is an example CI test pipeline that will run vulnerability scans on your image
```
Gotcha, will be in the next commit

Nava-JoshLong commented:
Github isn't letting me quote your overall comment, but I have them in an action since I wasn't able to run them in parallel due to not being able to share the image across jobs. I should've had that be the first line in my comment yesterday instead of in the end there. Since I can't share the image, I can't run the scans in parallel (and I haven't had a chance to just move them in the NJ repo since I'm playing catch-up this week)

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,81 @@
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          filters: |
+            needs-testing:
+              - app/**
+              - .github/**/*.yml
```
I bias towards simplicity, since I don't think the gain in runtime is worth the extra job and added dependency on a third party action. Since we're removing testing/linting from this workflow, I think it can be safe to say that if there are any changes in app/** we'd want to re-run scans, and specific projects can even narrow that down further e.g. if they want to narrow it to app/Dockerfile + app/src/**, but in template-infra we can't do that because we can't assume anything about the directory structure within app/

Nava-JoshLong commented:
@lorenyu The last outstanding thing is the failing check in the Dockle scan. Since this is the infra repo, I can mask the findings so that they aren't blocking the merge, or I can edit the Dockerfile to make the scan pass. Which do you prefer?

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,103 @@
+          exit-code: '1'
+          exit-level: WARN
+
+  image-scans:
```
Is this step actually needed? It looks like it's only purpose is to echo that the tests finished

Nava-JoshLong replied:
Technically, we don't but I had it there in case there was something that needed to be ran after the jobs finished. I'll remove it in the next commit since it isn't really needed

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,103 @@
+      - name: Build and tag Docker image for scanning
+        id: build-image
+        run: |
+          make release-build
+          echo "image=template-infra-app:latest" >> $GITHUB_OUTPUT
```
the image tag will only be `template-infra-app` in this template repo, but in a project repo it'll be `$PROJECT_NAME-$APP_NAME` where `PROJECT_NAME` defaults to the name of the repo directory, and `APP_NAME` defaults to `app` (you can see in the `Makefile` that `release-build` uses those environment variable `IMAGE_NAME` which is set to `$PROJECT_NAME-$APP_NAME`). So we might need to a way to get the actual `IMAGE_NAME` used from the `Makefile`. One idea is to add a `release-image-name` target to the `Makefile` like:

```make
release-image-name:
	@echo $(IMAGE_NAME)
```

then here we do something like
```suggestion
          make release-build
          IMAGE_NAME=$(make release-image-name)
          echo "image=$IMAGE_NAME:latest" >> $GITHUB_OUTPUT
```

not hundred percent of the `$(executable)` syntax works or if we need to put `make release-image-name` within backticks `, or do something like `make release-image-name > $IMAGE_NAME`

Nava-JoshLong replied:
Gotcha, adding to next commit

lorenyu reviewed (commented):
PR looks good to me! Just spotted one gotcha with the image name.

Regarding whether to ignore or fix the errors in the Dockerfile, how does ignoring the errors work? Can it be done in the Dockerfile itself or does it need to be done in the config files like .grype.yml and .trivyignore? If it can be done in the Dockerfile itself that'd be best so that we don't also cause projects to ignore their own issues when the copy over the config files. If not, then maybe we just need to bite the bullet and fix the issues in the template's example Dockerfile?

(If both options are equally easy for you, then I have a preference for editing the Dockerfile to make the scan pass)

Nava-JoshLong commented:
@lorenyu I fixed the issue in the Dockerfile from the scans, and changed from using the `latest` tag to the `IMAGE_TAG`, a la the `IMAGE_NAME` change you brought up. I also added a `help` command for the makefile that we can have make learning what things do a little easier

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,98 @@
+
+# NOTE: The workflow isn't able to pass the docker image between jobs, so each builds the image.
+#       A future PR will pass the image between the scans to reduce overhead and increase speed
+name: API Linting & Testing CI
```
looks like the name still references linting/testing, maybe rename to `CI Vulnerability Scans`

lorenyu commented on .github/workflows/ci-vulnerability-scans.yml:
```diff
@@ -0,0 +1,98 @@
+      - app/**
+  pull_request:
+    paths:
+        - app/**
```
nit, this line is indented differently
```suggestion
      - app/**
```

lorenyu commented on docs/infra/vulnerability-management.md:
```diff
@@ -0,0 +1,14 @@
+## How to use Workflow
+The workflow will run whenever there is a push to a PR or when merged to main if there are changes in the `app` direcotry. It is scanning in both cases to ensure there is no issues if a PR is approved on a Friday, but isn't merged till Monday - a CVE could have been found in the time between the last run and the merge.
+
+The job in the file, `image-scans`, is the orchestration job that will run the scans in parallel to speed up execution time. If there is a failure in any of the scans, this main job will fail as well
```
Looks like the doc still references the job that got removed. Maybe update or remove this paragraph?

lorenyu reviewed (approved):
LGTM! Just some final polish nits.

lorenyu commented on .dockleignore_acceptfiles:
```diff
@@ -0,0 +1,4 @@
+# This file is allows you to specify a list of files that is acceptable to Dockle
+# To allow multiple files, use a list of names. example below
+# DOCKLE_ACCEPT_FILES=file1,path/to/file2,file3/path,etc
+# https://github.com/goodwithtech/dockle#accept-suspicious-environment-variables--files--file-extensions
```
actually maybe a more precise name could be something like `.dockleconfig`

Nava-JoshLong replied:
Changed in latest push
